data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/consumer/UnorderedConsumerVerticle.java

```diff
@@ -81,11 +102,17 @@ private void poll() {
   }
 
   private void handleRecords(final KafkaConsumerRecords<Object, CloudEvent> records) {
-    final var futures = new ArrayList<Future>(records.size());
+    // We are not forcing the dispatcher to send less than `max.poll.records`
+    // requests because we don't want to keep records in-memory by waiting
+    // for responses.
     for (int i = 0; i < records.size(); i++) {
-      futures.add(this.recordDispatcher.dispatch(records.recordAt(i)));
+      this.inFlightRecords++;
+      this.recordDispatcher.dispatch(records.recordAt(i))
+        .onComplete(v -> {
+          this.inFlightRecords--;
+          poll();
+        });
     }
-    CompositeFuture.join(futures)
-      .onComplete(v -> poll());
+    poll();
```

Comment:
so we call the blocking `poll` now inner and outer loop, to ensure the inflight and max are in balance, before we call the actual consumer's poll?

Reply:
Yes, `poll` itself decides when to poll or wait for another call done by the handler in `onComplete()`.

nit: the `poll` is non-blocking.

Reply:
> itself decides when to poll or wait for 
ok, should we perhaps reflect that in the name of the method? `poll` seems very generic - but the method does more.

Sorry for the bikesheeding :sweat_smile:

Reply:
`maybePoll`? :smile: